archive/extract_activity.go

```diff
@@ -0,0 +1,177 @@
+package archive
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/google/safeopen"
+	"github.com/mholt/archiver/v4"
+	"go.artefactual.dev/tools/temporal"
+)
+
+const ExtractActivityName = "extract-archive-activity"
+
+var (
+	ErrNotAFile       = errors.New("Not a file")
+	ErrInvalidArchive = errors.New("Invalid archive")
+)
+
+type ExtractActivity struct {
+	cfg Config
+}
+
+func NewExtractActivity(cfg Config) *ExtractActivity {
+	return &ExtractActivity{cfg: cfg}
+}
+
+type ExtractActivityParams struct {
+	// SourcePath is the path of the archive to be extracted
+	SourcePath string
+}
+
+type ExtractActivityResult struct {
+	// DestPath is the path of the extracted archive contents.
+	DestPath string
+}
+
+// Execute extracts the content of SourcePath to a temporary directory, then
+// returns the DestPath of the extracted files.
+//
+// If SourcePath is a directory an ErrNotAFile error is returned. If SourcePath
+// is a file, but not a valid archive, an ErrInvalidArchive error is returned.
+func (a *ExtractActivity) Execute(
+	ctx context.Context, params *ExtractActivityParams,
+) (*ExtractActivityResult, error) {
+	logger := temporal.GetLogger(ctx)
+	logger.V(1).Info("Executing ExtractActivity", "SourcePath", params.SourcePath)
+
+	a.cfg.setDefaults()
+
+	dest, err := a.extract(ctx, params.SourcePath)
+	if err != nil {
+		switch err {
+		case ErrNotAFile:
+			logger.V(2).Info("ExtractActivity: not a file", "SourcePath", params.SourcePath)
+			return nil, err
+		case ErrInvalidArchive:
+			logger.V(2).Info("ExtractActivity: not a valid archive", "SourcePath", params.SourcePath)
+			return nil, err
+		default:
+			return nil, fmt.Errorf("extract archive: %v", err)
+		}
+	}
+
+	dest, err = skipTopLevelDir(dest)
+	if err != nil {
+		return nil, fmt.Errorf("extract archive: skipTopLevelDir: %v", err)
+	}
+
+	return &ExtractActivityResult{DestPath: dest}, nil
+}
+
+func (a *ExtractActivity) extract(ctx context.Context, src string) (string, error) {
+	fi, err := os.Stat(src)
+	if err != nil {
+		return "", err
+	}
+
+	if fi.IsDir() {
+		return src, ErrNotAFile
+	}
+
+	f, err := os.Open(src) // #nosec G304 -- trusted path.
+	if err != nil {
+		return "", fmt.Errorf("open: %v", err)
+	}
+	defer f.Close()
+
+	format, r, err := archiver.Identify(src, f)
+	if err != nil {
+		if errors.Is(err, archiver.ErrNoMatch) {
+			return "", ErrInvalidArchive
+		}
+		return "", fmt.Errorf("identify archive: %v", err)
+	}
+
+	var dest string
+	if ex, ok := format.(archiver.Extractor); ok {
+		dest, err = os.MkdirTemp("", "sdps-extract-archive")
```

Comment:
I'm having conflicts about the destination. I think we should accept a destination directory parameter and use it here, similar to what I did in the download activity in Enduro. If that parameter is empty, I'd consider placing the extracted content besides the initial file, instead of using the default temporary directory.

Reply:
I think adding a destination directory parameter is a good idea.  

I think using "/tmp" is a better default than writing the extracted content beside the initial file because:

1. I don't want to assume anything about the source directory. What if we don't have write permissions to it?
2. If the workflow fails after extraction and the destination directory is not cleaned up, then at least a reboot will delete the old extract directory. That said we should do our best to clean up the extract directory if something goes wrong, rather than waiting for a reboot.

In both cases (`dest` param and default "/tmp") I think we should use a unique extract directory in case the activity gets triggered twice on the same source archive somehow; we don't want to have two worker processes trying to write to the same destination.

Reply:
We may not have permissions to write to the temporary directory either, https://pkg.go.dev/os#TempDir. I'd like to keep as much as possible in a processing directory and not move things around and have the content scattered in the system, re https://github.com/artefactual-sdps/enduro/issues/926. Either way, the destination parameter is available to overcome those issues.

Reply:
Yeah, okay. You've convinced me. :)